client-gen/src/generator.rs

```diff
@@ -0,0 +1,459 @@
+use anyhow::Result;
+use convert_case::{Case, Casing};
+use sails_idlparser::ast::*;
+use std::path::{Path, PathBuf};
+
+use sails_idlparser::{ast::visitor, ast::visitor::Visitor};
+
+pub struct IdlGenerator {
+    path: PathBuf,
+}
+
+impl IdlGenerator {
+    pub fn new(path: PathBuf) -> Self {
+        Self { path }
+    }
+
+    pub fn generate(self, program: Program) -> Result<String> {
+        let mut trait_generator = RootGenerator::new("Service", &self.path);
+        visitor::accept_program(&program, &mut trait_generator);
+
+        let code = trait_generator.code;
+
+        let code = syn::parse_str(&code).expect("parse ast");
+        let code = prettyplease::unparse(&code);
+
+        Ok(code)
+    }
+}
+
+struct RootGenerator<'a> {
+    code: String,
+    service_name: &'a str,
+}
+
+impl<'a> RootGenerator<'a> {
+    fn new(service_name: &'a str, idl_path: &'a Path) -> Self {
+        let mut code = String::new();
+
+        code.push_str("// Code generated by sails-client-gen from ");
+        code.push_str(idl_path.file_name().unwrap().to_string_lossy().as_ref());
+        code.push_str(". DO NOT EDIT.\n");
+
+        code.push_str("#![automatically_derived]\n");
+        code.push_str("#![allow(unused)]\n");
+
+        code.push_str("use parity_scale_codec::{Encode, Decode};\n");
+        code.push_str("use sails_sender::Call;\n");
+        code.push_str("use gstd::prelude::*;\n");
+
+        Self { service_name, code }
+    }
+}
+
+impl<'a, 'ast> Visitor<'ast> for RootGenerator<'a> {
+    fn visit_service(&mut self, service: &'ast Service) {
+        let mut service_gen = ServiceGenerator::new(self.service_name.to_owned());
+        service_gen.visit_service(service);
+        self.code.push_str(&service_gen.code);
+
+        let mut structs_gen = RequestStructsGenerator::default();
+        structs_gen.visit_service(service);
+        self.code.push_str(&structs_gen.code);
+
+        let mut client_gen = ClientGenerator::new(self.service_name.to_owned());
+        client_gen.visit_service(service);
+        self.code.push_str(&client_gen.code);
+    }
+
+    fn visit_type(&mut self, t: &'ast Type) {
+        let mut type_generator = TypeGenerator::new(t.name());
+        type_generator.visit_type(t);
+        self.code.push_str(&type_generator.code);
+    }
+}
+
+#[derive(Default)]
+struct RequestStructsGenerator {
```

Comment:
You won't need these structs at all if you follow the suggested pseudo-code. You can always use tuples
```
let payload = [
    "MyFuncName".encode(),
    (p1, p2, p3, ...).encode()
].concat();
```

Reply:
rewritten